### ISSUE TITLE
Ignored private field still exported

### ISSUE THREAD

**Issue by jamesford42** [mono/CppSharp#748]
```
public unsafe partial class AttributeConfig : IDisposable
            {
                [StructLayout(LayoutKind.Explicit, Size = 32)]
                public partial struct __Internal
                {
                    [FieldOffset(0)]
                    internal global::Std.Vector.__Internal _attribs;
//....
```

Note that the _attribs is not actually used anywhere else in C#; it could totally not be there and everyone would be happy. It is there on the C++ type and needs to be, but it is private and marked CS_IGNORE which in previous versions meant it would not be emitted.

**Comment by tritao** [comment]
The field should only be generated on the internal struct. The problem here is that we need to have the P/Invoke struct fully qualified or in some cases the runtime might have trouble figuring out how to pass it to the native side regarding calling conventions. In some cases skipping them might work but it's better to be safe than sorry.

Since they are internal anyway, does it make any difference for your use case or is it just a cosmetic thing?

**Comment by jamesford42** [comment]
Yea, that wont compile, the internal field internal global::Std.Vector.__Internal _attribs; is not even a real type in my C# project. Neither is AttributeMetadata. Those are C++ system library types. The purpose of the AttributeConfig type (which I wrote) is to expose a what needs to be  used to C# without directly using complex system library types that cannot be exported.

**Comment by ddobrev** [comment]
@jamesford42 you need to include an additional generated file, it's called Std.cs. I would personally recommend an even more convenient solution, turn the option for automatic compilation on and you''ll directly get assemblies - unless we have generated invalid code because of bugs.

**Comment by jamesford42** [comment]
I had to revert to the version of CppSharp I was using before. I can't expose every single private C++ field to C#.

**Comment by ddobrev** [comment]
I am afraid I don't understand what the problem is. These fields are not visible in the public API.

**Comment by tritao** [comment]
@jamesford42 Can you explain the issue with the private fields? Is there some reason you cannot have them appear on the generated code? If we understand the issue, maybe we can figure out a solution that works for your needs.

**Comment by jamesford42** [comment]
They reference types that are unknown to my C# project and as a result do not build.

**Comment by ddobrev** [comment]
@jamesford42 please take a look at my explanation 5 comments above.

**Comment by jazzay** [comment]
I am having the same problem now that I have updated CppSharp. Previously private member variable pointer types were not generated at all. 

C++
```
class WrapperClass
{
public:
	void SomeMethod(int value);

private:
	CS_IGNORE class ClassNotWantedInCSharp* mField;
};
```

Generates this

C#
```
    public unsafe partial class ClassNotWantedInCSharp
    {
        [StructLayout(LayoutKind.Explicit, Size = 0)]
        public partial struct __Internal
        {
        }
    }

    public unsafe partial class WrapperClass: IDisposable
    {
        [StructLayout(LayoutKind.Explicit, Size = 12)]
        public partial struct __Internal
        {
            [FieldOffset(0)]
            public global::System.IntPtr mField;
        }
    }
```

Note that before I did not even need the CS_IGNORE macro. 

Is this something that can be fixed quickly? This new behavior is resulting in type collisions with unrelated Managed types.  As @jamesford42 mentions this is a pattern I use as well to hide complex native implementations that cannot be bound directly.

**Comment by ddobrev** [comment]
@jazzay as we told @jamesford42 , this is not a bug and is not going to be "fixed". These fields are only generated in the internal structure and serve marshalling purposes. In fact, in the summer we are going to need them even more because we are going to switch from explicit to sequential layouts. The latter require absolutely every field to be present since they do not use field offsets. From what you described the only piece which seems to be a bug is the type collision - I assume ambiguity - with real types. If that is the case, I'll be happy to continue this discussion.

**Comment by jazzay** [comment]
I still believe there is a bug here. The generated class (ClassNotWanted in example above) actually lives in the public namespace of the generated file. As you can see in my example it is not actually even used within the generated wrapping class - mField is an IntPtr, i do not want that exposed to bindings. In my Cpp header that is parsed this type is not exposed/included anywhere, it is forward declared and private - following the Impl pattern for C++. Private types should not be exposed to bindings for any reason imho. If you guys don't want to fix this then please point me in the right direction so I can make this fix in my own fork if necessary, which would be too bad.

**Comment by ddobrev** [comment]
@jazzay I agree classes the internals of which are only used through pointers as in your case can be removed. However, I have to ask you if such classes cause your generated code to be invalid. If they do, this is a bug. If not, it's a minor optimisation to be made. The proper solution depends on your answer.

**Comment by jazzay** [comment]
In my case I need to wrap the internal private native class with a native wrapper for binding, but I want the generated managed class to have the same name as the hidden C++ class for API reasons. So I used SetClassBindName to rename the binding wrapper. This worked fine before, but now with the hidden ptr class also being generated I get a conflict. The current inner native implementation is not easy to auto generate bindings for due to legacy dependencies, hence the binding wrapper. Hope this makes sense.

Also, it just feels/looks wrong exposing private implementation details to the public binding namespace.

**Comment by ddobrev** [comment]
@jazzay I understand. What you can do quickly is find the offending class in PostProcess and remove it from its parent, like this:

    var @class = astContext.FindClass("ClassNotWantedInCSharp");
    @class.Namespace.Declarations.Remove(@class);

The proper solution is a new pass for C++# which checks if classes with just internals are directly used for any field and if not, remove them. However, I cannot tell when we'd have the time to fix it ourselves so unless you'd like to try yourself, the workaround above should help you for the time being.
About exposing private implementation details, it is wrong. We intend to make all internals, well, internal, using the keyword of the same name. However, https://github.com/mono/CppSharp/issues/511 blocks us and we must solve that first.

**Comment by jazzay** [comment]
Thanks for the guidance. Will use that as a work around for now.
